### PR TITLE
Update OmniAuth to 2.0

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Dependency update

## Description

This PR

* updates OmniAuth related gems
* adds the [omniauth-rails_csrf_protection](https://github.com/cookpad/omniauth-rails_csrf_protection) gem as recommended in the [update docs](https://github.com/omniauth/omniauth/wiki/Upgrading-to-2.0#rails)
* removes CVE-2015-9284 from `bundle-audit`'s ignore list
* bundles Devise from the `0cd72a5` commit on the `master` branch. This is temporary, a new release should hopefully come soon.

## Related Tickets & Documents

Closes https://github.com/forem/forem/issues/12562

## QA Instructions, Screenshots, Recordings

Existing specs should be green

### UI accessibility concerns?

n/a

## Added tests?

- [X] No, and this is why: No new functionality

### CODE REVIEW

**Comment by citizen428** [comment]
@maestromac @fdxyz This is still in draft, just tagging you early so you can follow along.

**Comment by fdoxyz** [comment]
Awesome thank you for the heads up @citizen428!

Some things I think we'll probably need in order to make this fully work and remain secure would be:
- Remove our [CSRF workaround](https://github.com/forem/forem/blob/master/config/initializers/persistent_csrf_token_cookie.rb)
- Ensure our session cookie is `SameSite=None` and `Secure`
- Since we would be adding the [omniauth-rails_csrf_protection](https://github.com/cookpad/omniauth-rails_csrf_protection) gem then we no longer need the omniauth initializer code that [forced POST requests only](https://github.com/forem/forem/blob/master/config/initializers/omniauth.rb#L3-L7)

**Comment by maestromac** [comment]
So many vendor files :skull:

**Comment by rhymes** [comment]
@fdoxyz to do the first two steps we need to wait for the TLS work @maestromac is doing. `SameSite=none` should only be activated for the CSRF cookie, correct? Is that to make it work with Apple's OAuth?

**Comment by fdoxyz** [comment]
@rhymes what I originally envisioned was to remove the CSRF cookie altogether so we can stick with vanilla Rails CSRF protection in the session token.

I'm suggesting this because I ran into problems when doing this same attempt (using Devise custom branch install to test Omniauth 2.0). I believe these problems come from Omniauth's breaking changes in the 2.0 bump that rely on Rails core CSRF check, but I could be wrong.

Indeed, the SameSite=None is meant to solve the problem for Apple's OAuth, but I think it might be unsafe to enable this config if we are not relying on the Rails' core CSRF checks. Because of this I was suggesting to sort all of this out now that we're bumping Omniauth major version 🙂

This does have the side effect of possibly breaking some stuff because of the way we cache articles and our shell container (with the service worker). All of this to say I'm open to discussing our options and throughly think this through, but IMO sorting out these topics would be incredibly beneficial.

**Comment by rhymes** [comment]
@fdoxyz we can't really use "Standard Rails CSRF" protection but we might be able to get rid of the token.

See [Caching the Uncacheable: CSRF Security](https://www.fastly.com/blog/caching-uncacheable-csrf-security), we're using  "Technique 3: Token Over XHR"

**Comment by fdoxyz** [comment]
Thanks for clarifying @rhymes! I remember landing on that blog post from a link in an internal doc and possibly a previous conversation too; it still remains a tricky topic (at least for me lol)

I now looked it up and I see we are sending that token [here](https://github.com/forem/forem/blob/master/app/controllers/async_info_controller.rb#L5), so that technique's implementation ("Technique 3: Token Over XHR") and the FE code that relies on it doesn't need to change.

Based on the [initial 3 suggestions](https://github.com/forem/forem/pull/12715#issuecomment-780544777) I shared, we would only need 2 & 3 then. Both CSRF and session cookie need to be `SameSite=None` though, because [we need the session in the OAuth callback](https://github.com/fdoxyz/forem/blob/master/app/controllers/omniauth_callbacks_controller.rb#L55). If that's not a problem, then it should be okay to keep the extra cookie (the benefit is longer sessions if I'm not mistaken).

Anyways, I wanted to share these ideas because it would be great to sort out all our OAuth issues at once if possible. Looking forward to seeing this PR merged soon (when Devise release is available and possibly other blockers are ready too) 😃

**Comment by citizen428** [comment]
Thanks for the feedback/discussions so far, everyone!

* Regarding waiting for Devise to release, I think we could theoretically merge earlier by pinning to specific commit and not tracking `master`. It's not my preferred solution, but it would allow us to decouple this upgrade from Devise's release schedule.
* If I understand correctly, other fixes won't be possible for now as they still rely on TLS work, is this correct? In this case, I'd vote for getting this PR out of the way and putting a new card into the planning funnel for any follow-up work.

**Comment by citizen428** [comment]
I know pinned the Devise commit in the `Gemfile` and specs are green. Should we go ahead and merge this update first, then do a follow-up issue (ideally written by someone like @rhymes, who seems way more on top of this whole topic than I am)?

**Comment by citizen428** [comment]
One thing that slightly worries me: all specs here are green, but it's impossible to actually log in when starting a local server:

image

This may just be the nature of how we test things at the moment, but it's not particularly reassuring.

**Comment by rhymes** [comment]
I think I got lost a bit, regarding:

> Based on the initial 3 suggestions I shared, we would only need 2 & 3 then. Both CSRF and session cookie need to be SameSite=None though, because we need the session in the OAuth callback. If that's not a problem, then it should be okay to keep the extra cookie (the benefit is longer sessions if I'm not mistaken).

It's not super clear to me why we need `SameSite=None` though. Currently both DEV and Forem use `SameSite=Lax` and we can log in. Why do we need to switch to `None` ?

**Comment by fdoxyz** [comment]
> I'm currently looking into this, the problem seems to lie with OmniAuth::RequestForgeryProtection which includes ActionController::RequestForgeryProtection::TokenVerifier but does not have a cookies method, which leads to the NameError shown above.

This is in part what I meant when I mentioned the Rails vanilla CSRF checks (the `actionpack` code that protects from forgery), but I definitely have had trouble articulating what I meant by that since we have more than one custom technique at play here. It's quite a rabbit hole and I'm not sure I'm out of it yet 😅

`actionpack` uses `real_csrf_token(session)` to fetch the token for validation, but we are overriding it with our [CSRF workaround](https://github.com/forem/forem/blob/master/config/initializers/persistent_csrf_token_cookie.rb) and `cookies` isn't available in that context (when Omniauth is calling it [from this line](https://github.com/omniauth/omniauth/blob/master/lib/omniauth/strategy.rb#L236)).

Getting rid of our CSRF cookie workaround fixes this PR for me, or a patch like this also works:

```ruby
def real_csrf_token(session)
  if respond_to?(:cookies)
    csrf_token = cookies.encrypted[COOKIE_NAME] || session[:_csrf_token]
    csrf_token ||= SecureRandom.base64(AUTHENTICITY_TOKEN_LENGTH)
    cookies.encrypted[COOKIE_NAME] ||= {
      value: csrf_token,
      expires: 1.year.from_now,
      httponly: true
    }
    session[:_csrf_token] = csrf_token
    Base64.strict_decode64(csrf_token)
  else
    # Original implementation here: 
    # https://github.com/rails/rails/blob/v6.0.2/actionpack/lib/action_controller/metal/request_forgery_protection.rb#L396
    session[:_csrf_token] ||= SecureRandom.base64(AUTHENTICITY_TOKEN_LENGTH)
    Base64.strict_decode64(session[:_csrf_token])
  end
end
```

------------

> It's not super clear to me why we need SameSite=None though. Currently both DEV and Forem use SameSite=Lax and we can log in. Why do we need to switch to None ?

@rhymes after trying the above, I now think you're right about this and it's **not absolutely necessary**.

I originally thought it could've been related to the CSRF problems (not being able to reach the session), but it's true that **it will only continue to block Apple Auth**. Since we are focusing this work on revisiting Omniauth I thought we could take advantage of the opportunity and close the loop on it since we were revisiting all of these configs, but it may not be a blocker for bumping Omniauth.

------------

> Regarding waiting for Devise to release...

> ... Should we go ahead and merge this update first (after doing any necessary code changes of course), then do a follow-up issue...

I would be in favor of releasing with the pinned commit since we would be taking advantage of this extra CSRF validation check

**Comment by citizen428** [comment]
@fdoxyz Thanks for also diving into this and thanks for the workaround for the `cookies` issue. I added this know and after that I can indeed sign up, but then can't finish the onboarding process because of 

> ActionController::InvalidAuthenticityToken - ActionController::InvalidAuthenticityToken:
  app/middlewares/set_cookie_domain.rb:13:in `call'

which happens when hitting "Continue" here:

image

**Comment by citizen428** [comment]
@rhymes As discussed on Slack, I now added a DataDog counter for the invalid authenticity token errors. Comment posting is already handled in the frontend now:

image

I'm a bit unsure what else we should do in that `rescue_from` block, I spent the last hour trying different versions of `render` and `redirect_to`, but none of them worked reliably across different models and request types (normal POST, Ajax request).

**Comment by rhymes** [comment]
@citizen428 I think it's fine! I would also log the `request.path` (as a tag?), `controller_name` is not always self explanatory, and might require the operator to have to dig in the source code to figure out which page triggered the error.

**Comment by citizen428** [comment]
Good suggestion @rhymes, I added that now (I actually used `request.fullpath`, which includes the query string`).

File: Gemfile [hunk 1]
```diff
@@ -25,7 +25,7 @@ gem "carrierwave-bombshelter", "~> 0.2" # Protect your carrierwave from image bo
 gem "cloudinary", "~> 1.18" # Client library for easily using the Cloudinary service
 gem "counter_culture", "~> 2.7" # counter_culture provides turbo-charged counter caches that are kept up-to-date
 gem "ddtrace", "~> 0.45.0" # ddtrace is Datadog’s tracing client for Ruby.
-gem "devise", "~> 4.7" # Flexible authentication solution for Rails
+gem "devise", github: "heartcombo/devise", ref: "0cd72a5" # Flexible authentication solution for Rails
```
Comment (rhymes) [inline Gemfile:28]: we could add a note with https://github.com/forem/forem/pull/12715#issuecomment-781026686 explaining why we decide to merge with the unreleased devise

**Review by rhymes — COMMENTED** [review]
I tested with GitHub and Twitter, both sign in and out. 

It'd be great if we could test Facebook and Apple before merging this (cc @fdoxyz @joshpuetz)

One tiny detail, though I'm not 100% sure where we do render this page, the `devise/shared/_links.html.erb` file has slightly change to add `:post` to the `link_to`, which AFAIK won't work because we don't have `rails-ujs`, so we need to convert that into a `<form>`, similar to https://github.com/forem/forem/blob/master/app/views/shared/authentication/_providers_registration_form.html.erb

This is the part I'm referring to:

https://github.com/forem/forem/blob/e07f492221a5821a2994557a91bafb0275b6e575/app/views/devise/shared/_links.html.erb#L23

and this is how it is in Devise now 

https://github.com/heartcombo/devise/blob/0cd72a56f984a7ff089246f87a8b259120545edd/app/views/devise/shared/_links.html.erb#L23

File: Gemfile (hunk 1 above)
Comment (citizen428) [reply Gemfile:28]: 👍 , I'll add that tomorrow.

**Review by fdoxyz — APPROVED** [review]
I tested both Facebook and Apple. LGTM pending the comments from @rhymes above

> One tiny detail, though I'm not 100% sure where we do render this page, the devise/shared/_links.html.erb file has slightly changed to add :post to the link_to, which AFAIK won't work because we don't have rails-ujs, so we need to convert that into a...

The only place we are using `devise/shared/_links.html.erb` is from `/users/unlock` route, which is something email authenticated Forems most likely need but we haven't actually added proper crayons styling to it yet. 

It feels outside the scope of this PR to do work a lot on that page, but given that @rhymes is right (those links won't work and aren't even styled) my suggestion would be to just remove the render of the links [here](https://github.com/forem/forem/blob/master/app/views/devise/unlocks/new.html.erb#L16)

**Comment by citizen428** [comment]
@rhymes I went with @fdoxyz's suggestion here and just completely `devise/shared/_links.html.erb` from a page we clearly haven't put any styling effort into yet.

**Comment by rhymes** [comment]
@citizen428 @fdoxyz great idea! Maybe we can remove the `_links` partial as well

**Comment by citizen428** [comment]
I removed the partial now @rhymes

**Review by rhymes — APPROVED** [review]
LGTM!

**Comment by rhymes** [comment]
I'm waiting for the US team to be online in case we need to roll this back quickly. I want to be extra careful as it affects login

**Pull request merged**